### PR TITLE
Remove some unnecessary slicing from generated Regex code

### PR DESCRIPTION
**Pull request by stephentoub**
When we're outputting code to match a "multi" (a sequence of multiple characters), we're currently issuing a Slice for the known tracked offset even if that offset is 0.  We can skip that nop.

cc: @joperezr, @danmoseley 

<details>
<summary>Matching "defg" as part of a pattern before</summary>

```C#
// Multi "defg"
{
    byteSpan = global::System.Runtime.InteropServices.MemoryMarshal.AsBytes(textSpan);
    if ((uint)textSpan.Length < 4 ||
        global::System.Buffers.Binary.BinaryPrimitives.ReadUInt64LittleEndian(byteSpan.Slice(0)) != 0x67006600650064ul)
    {
        goto NoMatch;
    }
}
```
</details>

<details>
<summary>Matching "defg" as part of a pattern after</summary>

```C#
// Multi "defg"
{
    byteSpan = global::System.Runtime.InteropServices.MemoryMarshal.AsBytes(textSpan);
    if ((uint)textSpan.Length < 4 ||
        global::System.Buffers.Binary.BinaryPrimitives.ReadUInt64LittleEndian(byteSpan) != 0x67006600650064ul)
    {
        goto NoMatch;
    }
}
```
</details>

### CODE REVIEW

File: src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs [hunk 1]
```diff
@@ -1565,7 +1567,8 @@ void EmitOr()
                     // character-by-character while respecting the culture.
                     if (!caseInsensitive)
                     {
-                        using (EmitBlock(writer, $"if (!global::System.MemoryExtensions.StartsWith({textSpanLocal}.Slice({textSpanPos}), {Literal(node.Str)}))"))
+                        string sourceSpan = textSpanPos > 0 ? $"{textSpanLocal}.Slice({textSpanPos})" : textSpanLocal;
```
Comment (joperezr) [inline src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs:1570]: Can't we do this as well on the else block bellow when doing caseInsensitive?

**Review by joperezr — APPROVED** [review]
Other than the one comment this looks good. Is it worth adding a unit test that ensures the Slice is not there in this case? (if the tiny optimization doesn't warrant a test that's ok too)

File: src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs:1570]: You mean for the `{textSpanPos} + {i}`?  Yeah, that could be changed to be `{i}` if textSpanPos == 0.  It won't actually matter from a perf perspective as the `0 + i` will be simplified by the C# compiler to just `i`, but from a readability perspective it'd be nice to clean up. I'll do that, thanks.

**Review by bartonjs — APPROVED** [review]

**Comment by stephentoub** [comment]
> Other than the one comment this looks good. Is it worth adding a unit test that ensures the Slice is not there in this case?

This would likely manifest as a test which generated code and compared the generated output to a known string.  We currently don't have any such tests.  We could certainly add some (separately), though they would likely be a non-trivial time sink until the code has stabilized further, as many changes would likely require regenerating all such pre-generated snippets.

**Pull request merged**